lib/spack/spack/cmd/create.py

```diff
@@ -507,15 +524,19 @@ def get_name(args):
     # Default package name
     name = 'example'
 
-    if args.name:
+    if args.name is not None:
```

Comment:
Idk.  Are there cases?

Not providing the name argument defaults to `args.name` being `None`.  Providing `--name=""` in the original code would result in the name being taken from the URL, which is the same as not proving the `--name` option.

Providing `--name=""` with the recommended check (for `None`), results in an error:  `Error: Package name can only contain a-z, 0-9, and '-'`.

Reply:
If you aren't aware of cases where the distinction is important, IMO it's always safer to test `if var` vs. `if var is not None`, but

> Providing --name="" with the recommended check (for None)

indicates that you may not agree.

> Providing --name="" with the recommended check (for None), results in an error: Error: Package name can only contain a-z, 0-9, and '-'.

That there is an error is good. The message itself could be clearer. It looks like above there is logic to assign a default name if none is given:

```
    # Default package name
    name = 'example'
```

IMO it would be less confusing to remove that, and to fail if `args.name` is `""`.

Reply:
Hmm.  Most of the created packages are going to inherit from `Package` and the person creating the package is going to have to modify the generated template anyway.  So I'm not sure I see the harm in allowing the name to default to `example`. 

As for failing, the above check does fail with the provided error.  But I could error out earlier with a better message.

Reply:
> As for failing, the above check does fail with the provided error. But I could error out earlier with a better message.

That would be my preference because it clarifies how `""` is handled given the `if var is not None` check; with that additional check, `if var is not None` seems fine to me.

> I'm not sure I see the harm in allowing the name to default to example (except that the onus is on the user to also rename the associated repo directory).

Most importantly I don't think this behavior should necessarily change as part of this PR. I do think though that a user would never want a package named "example", and there may be cases where the default name would generate weird collisions if someone were to create several packages in a row and multiple did not have a detectable name.

Reply:
> > As for failing, the above check does fail with the provided error. But I could error out earlier with a better message.
> 
> That would be my preference because it clarifies how `""` is handled given the `if var is not None` check; with that additional check, `if var is not None` seems fine to me.
> 

I added the new error message.


> > I'm not sure I see the harm in allowing the name to default to example (except that the onus is on the user to also rename the associated repo directory).
> 
> Most importantly I don't think this behavior should necessarily change as part of this PR. I do think though that a user would never want a package named "example", and there may be cases where the default name would generate weird collisions if someone were to create several packages in a row and multiple did not have a detectable name.

I agree WRT creating an `example` package.  This PR has gotten relatively large and changing that "feature"  isn't exactly part of the original scope.